bjoern/_bjoernmodule.c

```diff
@@ -11,6 +11,10 @@ static PyObject*
 run(PyObject* self, PyObject* args)
 {
   ServerInfo info;
+#ifdef WANT_GRACEFUL_SHUTDOWN
+  info.active_connections = 0;
```

Comment:
Is there a built in way in libev to handle connection draining so that we don't have to keep a counter? I'm worried about counting bugs.

Reply:
Unfortunately no, going through the documentation I can't see any way to do that. In case it helps, I went through a couple of iterations of the counting code and this was by far the simplest one I came up with - counting connections as "active" once http-parser tells us a message has begun, and then not counting them once we're done with the sending the response.

I contemplated doing something like... storing all of the requests in an array/linked list/hash map on the ThreadInfo struct, and then basically watching for that to drop to either drop to zero entries, or if the existing entries aren't handling any connections. That would eliminate counting bugs, but it would also be significantly more complex as we'd need to handle growing the array occasionally if there's an influx of requests. I think we'd also need to store some additional state on the Request struct for whether or not it's currently handling a request. Given the complexity, that's why I decided to go for the simpler approach :)

That said though, I think it's worth doing some testing on dropping connections halfway through requests at various points in the lifecycle to see how the counting holds up. I'll go back and do that, and report on my findings :)

Reply:
My testing found something - aborted connections weren't being accounted for. I've fixed that up and added some tests for it as well.

Reply:
Why not implement counting as follows:

- A successful `accept` increments the counter
- A `close` decrements the counter

So, move to `ev_io_on_requests` and `close_connection`?

Reply:
We could do that! That would certainly be simpler. The downside to doing that is that any persistent connections that aren't being actively used will hold the event loop open until the timeout is reached. Is that acceptable?